### PR TITLE
Charuco Support

### PR DESCRIPTION
**Pull request by BytingBulldogs3539**
Add charuco calibration to photonvision. Currently does **not** support generating custom charuco boards. This does **not** support https://calib.io/pages/camera-calibration-pattern-generator. Currently only supports the 4X4_50 family. Also removes all dotboard calibration. Fixes using the lowest possible fps while doing calibration (now uses the highest fps available for each resolution).

### CODE REVIEW

File: photon-client/src/components/cameras/CameraCalibrationCard.vue [hunk 1]
```diff
@@ -5,59 +5,57 @@ import { CalibrationBoardTypes, type VideoFormat } from "@/types/SettingTypes";
...
+        ) *
+        (180 / Math.PI);
     }
+    uniqueResolutions.push(format);
   });
   uniqueResolutions.sort(
     (a, b) => b.resolution.width + b.resolution.height - (a.resolution.width + a.resolution.height)
   );
+
+  useStateStore().calibrationData.videoFormatIndex = uniqueResolutions[0].index;
```
Comment (mcm001) [inline photon-client/src/components/cameras/CameraCalibrationCard.vue:51]: What does adding this do? It seems odd that a getter would mutate this

File: photon-core/src/main/java/org/photonvision/common/util/TestUtils.java [hunk 2]
```diff
@@ -340,14 +340,14 @@ public static Path getPowercellImagePath(PowercellTestImages image, boolean test
...
-    public static Path getDotBoardImagesPath() {
-        return getResourcesFolderPath(false).resolve("calibrationBoardImages");
-    }
-
     public static Path getSquaresBoardImagesPath() {
         return getResourcesFolderPath(false).resolve("calibrationSquaresImg");
     }
 
+    public static Path getCharucoBoardImagesPath() {
+        return getResourcesFolderPath(false).resolve("calibrationCharucoImg");
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/common/util/TestUtils.java:348]: Is this still used?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/Calibrate3dPipe.java [hunk 3]
```diff
@@ -186,12 +188,16 @@ protected CameraCalibrationCoefficients calibrateMrcal(
         List<MatOfPoint2f> corner_locations =
                 in.stream().map(it -> it.imagePoints).map(MatOfPoint2f::new).collect(Collectors.toList());
 
+        List<MatOfFloat> levels =
+                in.stream().map(it -> it.levels).map(MatOfFloat::new).collect(Collectors.toList());
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/Calibrate3dPipe.java:192]: We create opencv mats here. Where do we release them?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/Calibrate3dPipe.java [hunk 4]
```diff
@@ -222,13 +228,36 @@ protected CameraCalibrationCoefficients calibrateMrcal(
...
             var tvec = new Mat();
+
+            Point3[] oPoints = o.objectPoints.toArray();
+            Point[] iPoints = o.imagePoints.toArray();
+
+            List<Point3> outputOPoints = new ArrayList<Point3>();
+            List<Point> outputIPoints = new ArrayList<Point>();
+
+            for (int i = 0; i < iPoints.length; i++) {
+                if (iPoints[i].x != -1 && iPoints[i].y != -1) {
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/Calibrate3dPipe.java:248]: Is comparing floats to ints
```suggestion
                if (iPoints[i].x != -1 && iPoints[i].y != -1) {
```
 here OK? Also generally confused what this code is doing. Please comment

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java [hunk 5]
```diff
@@ -219,16 +232,92 @@ private Size getWindowSize(MatOfPoint2f inPoints) {
...
+
+            if (!boardFound) {
+                // If we can't find a charuco board, give up
+                return null;
+            }
+
+            imgPoints.copyTo(outBoardCorners);
+            ObjPoints.copyTo(objPts);
+
+            if (params.useMrCal) {
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java:288]: Why does this work only need to happen for mrcal? Again document

File: photon-core/src/main/java/org/photonvision/common/util/TestUtils.java (hunk 2 above)
Comment (BytingBulldogs3539) [reply photon-core/src/main/java/org/photonvision/common/util/TestUtils.java:348]: Yes for unit tests.

**Comment by mcm001** [comment]
image
this is a new error with this, happens on page load. 
And when switching to charuco
image

File: photon-client/src/components/cameras/CameraCalibrationCard.vue [hunk 6]
```diff
@@ -53,11 +61,42 @@ const getUniqueVideoFormatsByResolution = (): VideoFormat[] => {
...
+//           calib.cameraIntrinsics.data[0]
+//         ) *
+//         (180 / Math.PI);
+//     }
+//     uniqueResolutions.push(format);
+//   });
+//   uniqueResolutions.sort(
+//     (a, b) => b.resolution.width + b.resolution.height - (a.resolution.width + a.resolution.height)
+//   );
+//   return uniqueResolutions;
```
Comment (mcm001) [inline photon-client/src/components/cameras/CameraCalibrationCard.vue:94]: stray commented out code

File: photon-client/src/types/WebsocketDataTypes.ts [hunk 7]
```diff
@@ -57,6 +57,7 @@ export interface WebsocketCameraSettingsUpdate {
   outputStreamPort: number;
   pipelineNicknames: string[];
   videoFormatList: WebsocketVideoFormat;
+  uniqueFormatList: WebsocketVideoFormat;
```
Comment (mcm001) [inline photon-client/src/types/WebsocketDataTypes.ts:60]: didnt we kill this?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java [hunk 8]
```diff
@@ -219,16 +232,94 @@ private Size getWindowSize(MatOfPoint2f inPoints) {
...
         // Convert the inFrame too grayscale to increase contrast
         Imgproc.cvtColor(inFrame, inFrame, Imgproc.COLOR_BGR2GRAY);
         boolean boardFound = false;
 
-        if (params.type == UICalibrationData.BoardType.CHESSBOARD) {
+        // Get the size of the inFrame
+        this.imageSize = new Size(inFrame.width(), inFrame.height());
+
+        if (params.type == UICalibrationData.BoardType.CHARUCOBOARD) {
+            Mat ObjPoints =
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java:252]: nit: variables should start lowercase

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java (hunk 5 above)
Comment (mcm001) [reply photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java:290]: this still seems smelly. Why should the choice of backend change the calibration detection result?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java [hunk 9]
```diff
@@ -242,60 +333,64 @@ private FindBoardCornersPipeResult findBoardCorners(Pair<Mat, Mat> in) {
...
         public FindCornersPipeParams(
                 int boardHeight,
                 int boardWidth,
                 UICalibrationData.BoardType type,
+                int tagFamily,
                 double gridSize,
-                FrameDivisor divisor) {
+                double markerSize,
+                FrameDivisor divisor,
+                boolean useMrCal) {
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/FindBoardCornersPipe.java:385]: same comment above, this needs to not care about choice of backend.

**Comment by mcm001** [comment]
So do we just not use the divisor at all when calibrating with charuco? If so, that should be hidden in the UI.

**Review by mcm001 — APPROVED** [review]

**Pull request merged**